### PR TITLE
Support multiple duplicate anilist ids in custom mappings

### PR DESCRIPTION
**Pull request by ghost**
Issue: [144](https://github.com/RickDB/PlexAniSync/issues/144)

This allows the usage of multiple anilist-ids duplicates with multiple seasons. By allowing this, long series such as fairy tail can have a set of plex seasons match one anilist item and another set match another anilist item.

This configuration is now supported:
```
seasons:
      - season: 1
        anilist-id: 6702
      - season: 2
        anilist-id: 6702
      - season: 3
        anilist-id: 20626
      - season: 4
        anilist-id: 20626
```

Testing data:
imagem

Results in Anilist:
imagem

### CODE REVIEW

**Comment by nm2325** [comment]
@reconman Not sure how the process is to get PRs approved to this repository.

Do you know if Rick does anything in this Repo. I'd like to at least updated the docker image since it's been stale for a while now. I'm guessing we would need help from @RickDB

**Comment by reconman** [comment]
I wanted to test it before merging but didn't have time yet because of Christmas.

**Comment by nm2325** [comment]
All good.
Sorry for intruding on your Christmas.

``` 
Btw I think the duplicate variable might not be entirely necessary since I ended up adding the seasons with certain id to the same object, so could check the length of that list instead of the duplicate parameter
```

File: anilist.py [hunk 1]
```diff
@@ -172,54 +172,78 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
+                            "duplicate_mappings": 0,
+                        })
+                        continue
+                    # For multiple seasons with the same anilist id check how they should be added together
+                    elif (index != 0 and plex_season.first_episode > plex_seasons[index - 1].last_episode):
+                        match["watched_episodes"] = plex_season.watched_episodes
+                    else:
+                        match["watched_episodes"] += plex_season.watched_episodes
+
+                        logger.warning(f"[MAPPING] Duplicate anilist-id found for custom mapping | "
```
Comment (reconman) [inline anilist.py:211]: I think we can remove this warning, otherwise it creates quite some spam for some series like One Piece.

File: anilist.py [hunk 2]
```diff
@@ -172,54 +172,78 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
+                    match = next(
+                        (item for item in anilist_matches if item['anilist_id'] == matched_id),
+                        None,
+                    )
+
+                    if not match:
+                        anilist_matches.append({
+                            "anilist_id": matched_id,
+                            "watched_episodes": plex_season.watched_episodes,
+                            "total_episodes": plex_season.last_episode,
```
Comment (reconman) [inline anilist.py:200]: total_episodes is never used and incremented incorrectly for the `elif (indes != 0` case

File: anilist.py [hunk 3]
```diff
@@ -172,54 +172,78 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
+                            "mapped_seasons": [plex_season.season_number],
+                            "duplicate_mappings": 0,
+                        })
+                        continue
+                    # For multiple seasons with the same anilist id check how they should be added together
+                    elif (index != 0 and plex_season.first_episode > plex_seasons[index - 1].last_episode):
```
Comment (reconman) [inline anilist.py:206]:
```suggestion
                    elif index > 0 and plex_season.first_episode > plex_seasons[index - 1].last_episode:
```

File: anilist.py [hunk 4]
```diff
@@ -172,54 +172,78 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
+            for match in anilist_matches:
+                logger.info(
+                    "[MAPPING] Custom Mapping of Title found | "
+                    f"title: {plex_title} | anilist id: {match['anilist_id']} | "
+                    f"total watched episodes: {match['watched_episodes']} | "
+                    f"seasons mapped: {match['mapped_seasons']} | "
                 )
+                # If we had custom mappings for multiple seasons with the same ID use
+                # cumulative episode count and skip per season processing
+                if match["duplicate_mappings"] == 0:
```
Comment (reconman) [inline anilist.py:232]: You're skipping the add_or_update for seasons where 1 season is mapped to 1 anilist id.

Bleach right now has 16 TVDB seasons mapped to anilist id 269 and 1 season to anilist id 116674. The author of the linked issue also has this situation with Fairy Tail season 8.

This loop doesn't update the episodes because of this condition, the loop below doesn't update because the season was already added to `custom_mapped_seasons`.

**Review by reconman — CHANGES_REQUESTED** [review]

File: anilist.py (hunk 2 above)
Comment (nm2325) [reply anilist.py:200]: Should have asked this. 
Should we allow the users to update a show on anilist if they don't have all the episodes on the previous season watched?
Like the example in the issue?

Should we create a configuration that always updates the anilist item with `watched episodes this season` + `number of episodes in previous seasons`? That was my intention with this last_episode parameter.

```
So with this
seasons:
      - season: 1
        anilist-id: 6702
        total_episodes: 25
        watched_episodes: 25
      - season: 2
        anilist-id: 6702
        total_episodes: 50
        watched_episodes: 10
      - season: 3
        anilist-id: 6702
        total_episodes: 50
        watched_episodes: 10
```

The calculations for episodes watched would be different. With the current setup we would always update `6702` with 40 episodes, do you think it makes sense to allow this script to update `6702` to 25+50+10 episodes?

File: anilist.py (hunk 4 above)
Comment (nm2325) [reply anilist.py:232]: Ah my bad.
I kept this condition because it was there originally but didn't really understand if there was something I was missing.
If this condition is not supposed to exist it actually makes the code a lot simpler.
Can remove the duplicate `parameter` entirely

File: anilist.py (hunk 2 above)
Comment (reconman) [reply anilist.py:200]: I think 85 is more correct than 45 (did you miscalculate?)

There are still issues with this approach when episodes are missing in previous seasons, but I can't think of any way to handle that.

Comment (nm2325) [reply anilist.py:200]: (Yes, I did)

Agree that 85 to me is more correct than 45 even if previous seasons have not been fully watched, I wouldn't understand why someone would skip seasons.

Regarding episodes from previous seasons being missing, doesn't seem really feasible so I guess our users will have to use the other configuration `start` to fix it. But we don't use this `start` here, I think that could be improved.

Comment (nm2325) [reply anilist.py:200]: I think I will create a new PR to add that functionality and keep this PR clean for now with the same behaviour

**Comment by nm2325** [comment]
Added support for `start` now and removed the first `elif` because it actually wasn't doing what we wanted.

If we had a season with `25` episodes and then one season with `50` mapped to the same id. If the second season had more than 25 episodes watched, then anilist would get only those `25+` and never add them together to have `50+` episodes updated for that season.

**Comment by nm-remarkable** [comment]
This is my account now. Makes it easier to only have one account

File: anilist.py [hunk 5]
```diff
@@ -832,7 +819,7 @@ def map_watchcount_to_seasons(title: str, season_mappings: List[AnilistCustomMap
 
     for mapping in season_mappings:
         if watched_episodes >= mapping.start:
-            episodes_in_season = (watched_episodes - mapping.start + 1)
+            episodes_in_season = (watched_episodes - mapping.start)
```
Comment (reconman) [inline anilist.py:822]: If we have a custom mapping where Anilist season 2 begins at Plex episode 14, the mapping would have `start: 14` in it.

If the user watches episode 14, `episodes_in_season` would be set to `0` and then nothing gets updated.

Please revert this change and the one in custom_mappings.py.

File: anilist.py [hunk 6]
```diff
@@ -194,43 +195,29 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
                             "mapped_seasons": [plex_season.season_number],
-                            "duplicate_mappings": 0,
                         })
                         continue
-                    # For multiple seasons with the same anilist id check how they should be added together
-                    elif (index != 0 and plex_season.first_episode > plex_seasons[index - 1].last_episode):
-                        match["watched_episodes"] = plex_season.watched_episodes
+                    # For multiple seasons with the same id
+                    # If the start of this season has been mapped use that
+                    if mapped_start:
```
Comment (reconman) [inline anilist.py:207]:
```suggestion
                    if mapped_start > 0:
```

Integers with value 0 are inherently `false`, but I'd rather write it like this for readability.

Also with your current branch, the `if` is always entered since custom_mappings.py sets `start` to `1` if `start` was not specified. `1` is `true`, that's why the else is never called.

File: anilist.py [hunk 7]
```diff
@@ -194,43 +195,29 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
-                                       f"title: {plex_title} | "
-                                       f"anilist id: {matched_id} | "
-                                       f"total watched episodes: {match['watched_episodes']} | "
-                                       f"previous season matches: {match['mapped_seasons']} | "
-                                       f"current season: {plex_season.season_number} |"
-                                       )
-
-                    match["total_episodes"] = + plex_season.last_episode
+                    # TODO support using number of last episode of the last season as a start
                     match["mapped_seasons"].append(plex_season.season_number)
```
Comment (reconman) [inline anilist.py:213]: mapped_seasons is never read

**Review by reconman — CHANGES_REQUESTED** [review]

File: anilist.py (hunk 6 above)
Comment (nm-remarkable) [reply anilist.py]: It is a bit confusing to do this mapped_start when the default is 1.
I think I got it right now.

File: anilist.py (hunk 7 above)
Comment (nm-remarkable) [reply anilist.py:214]: My itentention was to add it to the logger and present the List of seasons that matched with the same AnilistID.
Have added it to the logger instead of removing it, I think it would be beneficial for the users.

**Comment by nm-remarkable** [comment]
I ask you once again, when you have the time to review these changes 🙏

File: anilist.py [hunk 8]
```diff
@@ -172,54 +172,67 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
+                    match = next(
+                        (item for item in anilist_matches if item['anilist_id'] == matched_id),
+                        None,
+                    )
+
+                    if not match:
+                        # Create first match dict for this anilist id
+                        anilist_matches.append({
+                            "anilist_id": matched_id,
+                            "watched_episodes": mapped_start + plex_season.watched_episodes,
```
Comment (reconman) [inline anilist.py:201]:
```suggestion
                            "watched_episodes": plex_season.watched_episodes - mapped_start + 1,
```

When I tested it, it updated some Season 2+ entries with Plex episode count + 1.

**Review by reconman — CHANGES_REQUESTED** [review]

File: anilist.py [hunk 9]
```diff
@@ -172,54 +172,67 @@ def match_to_plex(anilist_series: List[AnilistSeries], plex_series_watched: List
...
+                            "total_episodes": plex_season.last_episode,
+                            "mapped_seasons": [plex_season.season_number],
+                        })
+                        continue
+                    # For multiple seasons with the same id
+                    # If the start of this season has been mapped use that.
+                    if mapped_start > 1:
```
Comment (reconman) [inline anilist.py:208]:
```suggestion
                    if mapped_start != 1:
```
With this change, I could track my One Piece episodes when the previous seasons are missing.

I just have to use a `start` value of `2 - current_season.first_episode`.

**Review by reconman — CHANGES_REQUESTED** [review]

**Comment by reconman** [comment]
Closing because I can't edit your PR branch, continuing in #174
